### ISSUE TITLE
Adding a test with openssl or variants without TLS v1.3 support

### ISSUE THREAD

**Issue by veblush** [grpc/grpc#24960]
From the recent regression fixed by #24955, it's necessary to have a test with openssl or variants without TLS v1.3 support. It will be added to Linux primarily because it's quite easy to establish those environment with docker. Libressl could be considered to be added too.

**Comment by jiangtaoli2016** [comment]
Another related issue https://github.com/grpc/grpc/pull/24959

**Comment by matthewstevenson88** [comment]
As @jiangtaoli2016 mentioned, it would be useful to also test OpenSSL 1.0 and BoringSSL 1.0, since SSL transport security has a bunch of #ifdef's that single out those older versions.

**Comment by veblush** [comment]
I'm trying to understand what's needed specifically to prevent the previous bug. I like to know **what test** could've run on **what environment**. Is there any existing unit test to run for this purpose? Also if we managed to run some test for this, if client and server are using the same OpenSSL 1.0 which doesn't support TLS1.3, would it run into the error? Does it need an environment where the client is built with openssl 1.0 but the server is built with openssl 1.1?

**Comment by matthewstevenson88** [comment]
Thanks for looking into this Esun! The specific failure that was fixed by #24995 would have definitely been caught by an interop test where e.g. the client is using LibreSSL and the server is using a newer OpenSSL version. Here, client/server should be interchangeable, it would have been caught if the client was using the new OpenSSL version and the server was using LibreSSL, I think.

If the client and server were both running OpenSSL 1.0 (or one was running OpenSSL 1.0 and the other was running a newer OpenSSL version), then we would **not** have seen this exact issue, I think. I suggested testing the older OpenSSL versions because I'm worried about preventing a similar bug in the future.

To guard against future bugs, the ideal scenario would be to have the following tests:

- Run [ssl_transport_security_test.cc](https://github.com/grpc/grpc/blob/master/test/core/tsi/ssl_transport_security_test.cc), [ssl_session_cache_test.cc](https://github.com/grpc/grpc/blob/master/test/core/tsi/ssl_session_cache_test.cc), and the C-core e2e test with `h2_ssl_*` or `h2_tls` [fixtures](https://github.com/grpc/grpc/tree/master/test/core/end2end/fixtures) against new/old OpenSSL, new/old BoringSSL, LibreSSL.
- In the interop tests, we start a C++ client built with SSL library X and a C++ server built with SSL library Y, and X,Y range over new/old OpenSSL, new/old BoringSSL, LibreSSL.

I know that the above is a long wish list and might not be easily done, so I'm happy to help triage and figure out what is most important. :)

**Comment by veblush** [comment]
@matthewstevenson88 Thank you for detailed information! I briefly checked the current test harness and it doesn't seem easy to add new tests like that. (One of difficulty is that our interop test is using bazel build which supports boringssl only) So one of the simplest solution is adding portability test using old version of openssl (e.g. 1.0) and running unittests and e2e tests you mentioned. Do you think this would be helpful?

**Comment by matthewstevenson88** [comment]
Yes, that would be super helpful! Thanks Esun! Once that's done, let me know and, if needed, I can add variants of the unit tests and e2e tests to make sure we're getting good coverage of the codepaths that depend on the OpenSSL version.

Also, if the interop tests in my previous comment are not feasible, is it possible to add a portability test using LibreSSL (the most recent version)?

I don't know if gRPC officially supports this SSL library, but having this portability test is valuable for a few reasons: it was the SSL library that triggered the bug that sparked this discussion, it is the default SSL library on newer Macs, and its TLS semantics are slightly different from the newer OpenSSL/BoringSSL versions.

Please feel free to push back on this if I'm signing you up for too much work - I don't know how onerous it is to add a new portability test.

**Comment by veblush** [comment]
@matthewstevenson88

From the [fusion](https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod:grpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_portability), it appears that newly added test started failing. Can you take a quick look at this error? If you think it's easy to fix, we can try it without getting it rolled back. Otherwise I should revert it.

**Comment by matthewstevenson88** [comment]
From a quick look at the failures, it seems there are at least 2 different problems, so I think we should revert it for now.

I'm happy to help debug - is there a way to run this portability test locally? Or do I need to use RBE? If I need to use RBE, what should I set the `--bazelrc=` flag to in order to run this test?

**Comment by veblush** [comment]
@matthewstevenson88 Thank you for the quick response. I just made https://github.com/grpc/grpc/pull/25812 to roll it back. For the reproduction, let me try to get a command to run the test. This isn't possible to run with bazel rbe so I need to do some manual approach.

**Comment by veblush** [comment]
### How to reproduce it

```
# On grpc-repo
tools/run_tests/start_port_server.py
docker run -it --rm --net=host grpctesting/cxx_buster_openssl102_x64:0b42f85ef7d64afd880c9ca8acbc602fbc003cd6

# Clone
git clone https://github.com/grpc/grpc
cd grpc
git submodule update --init

# Build
mkdir -p cmake/build && cd cmake/build
cmake -DgRPC_BUILD_TESTS=ON -DgRPC_SSL_PROVIDER=package ../..
make -j server_ssl_test client_ssl_test ssl_transport_security_test
cd ../..

# Run tests
export GRPC_VERBOSITY=DEBUG
cmake/build/server_ssl_test
cmake/build/client_ssl_test
cmake/build/ssl_transport_security_test 
```

**Comment by matthewstevenson88** [comment]
Thanks @veblush! I'll try to sort out the fixes this week.

**Comment by matthewstevenson88** [comment]
It turns out there are quite a few bugs and test failures. I've fixed the first set of them in PR #25843, which allows ssl_transport_security_test.cc to pass when built against OpenSSL 1.0.2. Will look at the other test failures later this week.

**Comment by matthewstevenson88** [comment]
One additional thought: when #25860 is merged (and assuming there are no further errors), we can officially advertise that gRPC C-core (and wrapped languages) support OpenSSL 1.0.2. This is notable because OpenSSL 1.0.2 is the only FIPS approved version at present, I think.
